sql/core/src/main/scala/org/apache/spark/sql/dynamicpruning/PartitionPruning.scala

```diff
@@ -0,0 +1,264 @@
+    case Not(expr) => isLikelySelective(expr)
+    case And(l, r) => isLikelySelective(l) || isLikelySelective(r)
+    case Or(l, r) => isLikelySelective(l) && isLikelySelective(r)
+    case Like(_, _) => true
+    case _: BinaryComparison => true
+    case _: In | _: InSet => true
+    case _: StringPredicate => true
+    case _ => false
+  }
+
+  /**
+   * Search a filtering predicate in a given logical plan
+   */
+  private def hasSelectivePredicate(plan: LogicalPlan): Boolean = {
+    plan.find {
+      case f: Filter => isLikelySelective(f.condition)
+      case _ => false
+    }.isDefined
+  }
+
+  /**
+   * To be able to prune partitions on a join key, the filtering side needs to
+   * meet the following requirements:
+   *   (1) it can not be a stream
+   *   (2) it needs to contain a selective predicate used for filtering
+   *   (3) the join key is in the attribute set of the filtering plan
+   */
+  private def hasPartitionPruningFilter(key: Expression, plan: LogicalPlan): Boolean = {
+    !plan.isStreaming && hasSelectivePredicate(plan) &&
+      key.references.subsetOf(plan.outputSet)
+  }
+
+  private def canPruneLeft(joinType: JoinType): Boolean = joinType match {
+    case Inner | LeftSemi | RightOuter => true
+    case _ => false
+  }
+
+  private def canPruneRight(joinType: JoinType): Boolean = joinType match {
+    case Inner | LeftOuter => true
+    case _ => false
+  }
+
+  private def prune(plan: LogicalPlan): LogicalPlan = {
+    plan transformUp {
+      // skip this rule if there's already a DPP subquery on the LHS of a join
+      case j @ Join(Filter(_: DynamicPruningSubquery, _), _, _, _, _) => j
+      case j @ Join(_, Filter(_: DynamicPruningSubquery, _), _, _, _) => j
+      case j @ Join(left, right, joinType, Some(condition), hint) =>
+        var newLeft = left
+        var newRight = right
+
+        // extract the left and right keys of the join condition
+        val (leftKeys, rightKeys) = j match {
+          case ExtractEquiJoinKeys(_, lkeys, rkeys, _, _, _, _) => (lkeys, rkeys)
+          case _ => (Nil, Nil)
+        }
+
+        // checks if two expressions are on opposite sides of the join
+        def fromDifferentSides(x: Expression, y: Expression): Boolean = {
+          def fromLeftRight(x: Expression, y: Expression) =
+            !x.references.isEmpty && x.references.subsetOf(left.outputSet) &&
+              !y.references.isEmpty && y.references.subsetOf(right.outputSet)
+          fromLeftRight(x, y) || fromLeftRight(y, x)
+        }
+
+        splitConjunctivePredicates(condition).foreach {
+          case EqualTo(a: Expression, b: Expression)
+              if fromDifferentSides(a, b) =>
+            val (l, r) = if (a.references.subsetOf(left.outputSet) &&
+              b.references.subsetOf(right.outputSet)) {
+              a -> b
+            } else {
+              b -> a
+            }
+
+            // there should be a partitioned table and a filter on the dimension table,
```

Comment:
> ... and a filter on the dimension table

what if the dimension table itself is very small?

Reply:
If the dimension table is smaller than the table(s) on build side, the `pruningHasBenefit` will return false, which means:
1) the filter can still be planned as a broadcast pruning filter if it's eventually a BHJ with the expected build plan here as the build side, and that broadcast will be a reuse so hopefully not much overhead anyway;
2) otherwise, since `pruningHasBenefit` is false, this filter will be turned into a bypass filter (constant `true`), so nothing will happen.

Reply:
> the filter can still be planned as a broadcast pruning filter if it's eventually a BHJ

But from the code, we only add the filter if `hasPartitionPruningFilter(r, right)` is true, which requires a seletive predicate on the build side.

Reply:
We can address it in a followup. The current check is more conservative.